## 1. Imporing Libraries

In [12]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

## 2. Reading Data & Preliminary Check

In [20]:
flight = pd.read_csv('../data/flight_price.csv')
flight

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [21]:
flight.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='str')

In [22]:
flight.info()

<class 'pandas.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   Airline          10683 non-null  str  
 1   Date_of_Journey  10683 non-null  str  
 2   Source           10683 non-null  str  
 3   Destination      10683 non-null  str  
 4   Route            10682 non-null  str  
 5   Dep_Time         10683 non-null  str  
 6   Arrival_Time     10683 non-null  str  
 7   Duration         10683 non-null  str  
 8   Total_Stops      10682 non-null  str  
 9   Additional_Info  10683 non-null  str  
 10  Price            10683 non-null  int64
dtypes: int64(1), str(10)
memory usage: 918.2 KB


In [52]:
flight.duplicated().sum()

np.int64(220)

In [55]:
flight.loc[flight.duplicated(keep=False)].sort_values(['Airline','Date_of_Journey','Source','Destination'])

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

### Airline

In [107]:
flight.Airline.unique()

<StringArray>
[                           'IndiGo',                         'Air India',
                       'Jet Airways',                          'SpiceJet',
                 'Multiple carriers',                             'GoAir',
                           'Vistara',                          'Air Asia',
           'Vistara Premium economy',              'Jet Airways Business',
 'Multiple carriers Premium economy',                            'Trujet']
Length: 12, dtype: str

### Date_of_Journey

In [114]:
flight.Date_of_Journey.unique()

<StringArray>
['24/03/2019',  '1/05/2019',  '9/06/2019', '12/05/2019', '01/03/2019',
 '24/06/2019', '12/03/2019', '27/05/2019',  '1/06/2019', '18/04/2019',
  '9/05/2019', '24/04/2019',  '3/03/2019', '15/04/2019', '12/06/2019',
  '6/03/2019', '21/03/2019',  '3/04/2019',  '6/05/2019', '15/05/2019',
 '18/06/2019', '15/06/2019',  '6/04/2019', '18/05/2019', '27/06/2019',
 '21/05/2019', '06/03/2019',  '3/06/2019', '15/03/2019',  '3/05/2019',
  '9/03/2019',  '6/06/2019', '24/05/2019', '09/03/2019',  '1/04/2019',
 '21/04/2019', '21/06/2019', '27/03/2019', '18/03/2019', '12/04/2019',
  '9/04/2019',  '1/03/2019', '03/03/2019', '27/04/2019']
Length: 44, dtype: str

### Source

In [123]:
flight.Source.unique()

<StringArray>
['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai']
Length: 5, dtype: str

### Destination

In [125]:
flight.Destination.unique()

<StringArray>
['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad']
Length: 6, dtype: str

### Dep_Time

In [140]:
flight.Dep_Time.loc[lambda x:x.str.contains("[^0-9:]",regex=True)]

Series([], Name: Dep_Time, dtype: str)

### Arrival_Time

In [158]:
flight.Arrival_Time.loc[lambda x:x.str.contains("[^0-9:]",regex=True)].str.split(' ',n=1).str.get(1).unique()

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

### Duration

In [168]:
flight.Duration.loc[lambda x: ~ x.str.contains('m')].unique()

<StringArray>
['19h', '23h', '22h', '12h',  '3h',  '5h', '10h', '18h', '24h', '15h', '16h',
  '8h', '14h', '20h', '13h', '11h',  '9h', '27h', '26h',  '4h',  '7h', '30h',
 '21h', '28h', '47h',  '6h', '25h', '38h', '34h']
Length: 29, dtype: str

In [171]:
flight.Duration.loc[lambda x: ~x.str.contains('h')]

6474    5m
Name: Duration, dtype: str

In [186]:
flight.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

In [212]:
(
    flight.drop(index=6474).Duration.str.split(' ',expand=True).set_axis(['hour','minutes'],axis=1)
    .assign(
        hour = lambda x:x.hour.str.replace('h','').astype('int').mul(60),
        minutes = lambda x:x.minutes.str.replace('m','').fillna('0').astype('int')
    )
    .sum(axis=1)
    .rename('duration_minutes')
    .to_frame()
    .join(flight.drop(index=6474).Duration)
)

duration_minutes Duration
0                   170   2h 50m
1                   445   7h 25m
2                  1140      19h
3                   325   5h 25m
4                   285   4h 45m
...                 ...      ...
10678               150   2h 30m
10679               155   2h 35m
10680               180       3h
10681               160   2h 40m
10682               500   8h 20m

[10682 rows x 2 columns]

### Total_Stops

In [229]:
flight.Total_Stops.str.replace('non-stop','0 stops').str.replace('stops?','',regex=True).unique()

<StringArray>
['0 ', '2 ', '1 ', '3 ', nan, '4 ']
Length: 6, dtype: str

### Additional_Info

In [234]:
flight.Additional_Info.str.replace('No Info','No info').unique()

<StringArray>
[                     'No info',  'In-flight meal not included',
 'No check-in baggage included',              '1 Short layover',
               '1 Long layover',              'Change airports',
               'Business class',               'Red-eye flight',
               '2 Long layover']
Length: 9, dtype: str

## 3. Observation
- `Route` & `Total_Stops` columns are having missing value.
- `Date_of_Journey`, `Dep_Time` & `Arrival_Time` columns are having wrong datatypes. They should be in datetime format.
- `Duration` & `Total_Stops` columns can be integers.
- There are 220 records as duplicates.
- `Airline` column is having is having seat category like Premium economy, Business. They need to be removed. Should be in Title format.
- `Destination` column is having New Delhi and Delhi. It should be cleaned.
- `Route` column can be deleted as it is not giving much information.

## 4. Cleaning

In [216]:
def handle_duration(ser):
    return (
        ser
        .duration.str.split(' ',expand=True).set_axis(['hour','minutes'],axis=1)
        .assign(
            hour = lambda x:x.hour.str.replace('h','').astype('int').mul(60),
            minutes = lambda x:x.minutes.str.replace('m','').fillna('0').astype('int')
        )
        .sum(axis=1)
    )

In [243]:
def clean_data(df):
    return (df
        .drop(index=6474)
        .drop_duplicates()
        .assign(**{
            col: df[col].str.strip() for col in df.select_dtypes(include=['string']).columns
                })
        .rename(columns=str.lower)
        .assign(
            airline = lambda x:x.airline.str.replace(" Premium economy","").str.replace(" Business","").str.title(),
            date_of_journey = lambda x:pd.to_datetime(x.date_of_journey,dayfirst=True),
            destination = lambda x:x.destination.str.replace("New ",""),
            dep_time = lambda x:pd.to_datetime(x.dep_time).dt.time,
            arrival_time = lambda x:pd.to_datetime(x.arrival_time).dt.time,
            duration = lambda x:x.pipe(handle_duration),
            total_stops = lambda x:x.total_stops.str.replace('non-stop','0 stops').str.replace('stops?','',regex=True).astype('float'),
            additional_info = lambda x:x.additional_info.str.replace('No Info','No info')
        )
        .drop(columns=['route'])
    )

cleaned_flight = clean_data(flight)
cleaned_flight

/var/folders/c3/qlc29pfs0cg2d8j3tpcgz2m00000gp/T/ipykernel_85052/3818662244.py:13: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time = lambda x:pd.to_datetime(x.dep_time).dt.time,
/var/folders/c3/qlc29pfs0cg2d8j3tpcgz2m00000gp/T/ipykernel_85052/3818662244.py:14: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time = lambda x:pd.to_datetime(x.arrival_time).dt.time,


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore       Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore       Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore       Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No info   3897  
1         13:15:00       445          2.0         No info   7662  
2         04:25:00      1140          2.0         No info  13882  
3         23:30:00       325          1.0         No info   6218  
4         21:35:00       285          1.0         No info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No info   4107  
10679     23:20:00       155          0.0         No info   4145  
10680     11:20:00       180          0.0         No info   7229  
10681     14:10:00       160          0.0         No info  12648  
10682     19:15:00       500          2.0         No info  11753  

[10462 rows x 10 columns]

## 5. Splitting the Dataset

In [244]:
X = cleaned_flight.drop(columns=['price'])
y = cleaned_flight.price

In [245]:
X_,X_test,y_,y_test = train_test_split(X,y,test_size=0.2,random_state=42)
X_train,X_val,y_train,y_val = train_test_split(X_,y_,test_size=0.2,random_state=42)

In [1]:
def export_data(x,y,filename):
    final_df = x.join(y)
    final_df.to_csv('../data/'+filename+'_data.csv',index=False)
    return final_df

In [262]:
export_data(X_train,y_train,'train')

airline date_of_journey    source destination  dep_time  \
3635  Jet Airways      2019-03-21  Banglore       Delhi  08:55:00   
613   Jet Airways      2019-03-27     Delhi      Cochin  17:30:00   
7390        Goair      2019-03-09  Banglore       Delhi  11:40:00   
7575    Air India      2019-06-12   Kolkata    Banglore  09:25:00   
7032  Jet Airways      2019-03-12  Banglore       Delhi  22:55:00   
...           ...             ...       ...         ...       ...   
4128  Jet Airways      2019-03-21     Delhi      Cochin  10:45:00   
2836    Air India      2019-05-01   Kolkata    Banglore  09:25:00   
6664  Jet Airways      2019-06-01     Delhi      Cochin  14:00:00   
7999     Air Asia      2019-06-24     Delhi      Cochin  07:55:00   
6024    Air India      2019-03-01  Banglore       Delhi  11:50:00   

     arrival_time  duration  total_stops              additional_info  price  
3635     19:10:00       615          1.0  In-flight meal not included   7832  
613      04:25:00       655          1.0  In-flight meal not included   6540  
7390     14:35:00       175          0.0                      No info   7305  
7575     18:30:00       545          1.0                      No info   8366  
7032     07:40:00       525          1.0  In-flight meal not included  11087  
...           ...       ...          ...                          ...    ...  
4128     18:50:00      1925          2.0                      No info  11093  
2836     18:30:00       545          1.0                      No info   8891  
6664     19:00:00       300          1.0  In-flight meal not included  10262  
7999     13:25:00       330          1.0                      No info   6152  
6024     08:55:00      1265          1.0               1 Long layover  14594  

[6695 rows x 10 columns]

In [263]:
export_data(X_test,y_test,'test')

airline date_of_journey    source destination  dep_time  \
2150         Jet Airways      2019-03-06  Banglore       Delhi  08:00:00   
3784            Spicejet      2019-06-06   Kolkata    Banglore  22:20:00   
714               Indigo      2019-03-18   Kolkata    Banglore  05:30:00   
7559              Indigo      2019-06-27   Chennai     Kolkata  19:35:00   
7414              Indigo      2019-05-06   Kolkata    Banglore  15:15:00   
...                  ...             ...       ...         ...       ...   
653          Jet Airways      2019-05-27     Delhi      Cochin  19:15:00   
8608   Multiple Carriers      2019-06-27     Delhi      Cochin  11:25:00   
10261        Jet Airways      2019-06-03     Delhi      Cochin  02:15:00   
10581  Multiple Carriers      2019-06-06     Delhi      Cochin  15:15:00   
5285   Multiple Carriers      2019-06-03     Delhi      Cochin  18:15:00   

      arrival_time  duration  total_stops              additional_info  price  
2150      08:15:00      1455          1.0                      No info  17996  
3784      00:40:00       140          0.0                      No info   3873  
714       08:20:00       170          0.0                      No info   4462  
7559      21:55:00       140          0.0                      No info   3597  
7414      17:45:00       150          0.0                      No info   4804  
...            ...       ...          ...                          ...    ...  
653       12:35:00      1040          1.0  In-flight meal not included  12898  
8608      19:15:00       470          1.0                      No info   7155  
10261     04:25:00      1570          1.0  In-flight meal not included  11627  
10581     01:30:00       615          1.0                      No info   6795  
5285      01:30:00       435          1.0                      No info  13377  

[2093 rows x 10 columns]

In [264]:
export_data(X_val,y_val,'val')

airline date_of_journey    source destination  dep_time  \
5617             Indigo      2019-06-24     Delhi      Cochin  20:25:00   
6228  Multiple Carriers      2019-06-12     Delhi      Cochin  09:45:00   
6772        Jet Airways      2019-03-12  Banglore       Delhi  22:55:00   
6423  Multiple Carriers      2019-06-06     Delhi      Cochin  13:00:00   
1639        Jet Airways      2019-05-18     Delhi      Cochin  23:05:00   
...                 ...             ...       ...         ...       ...   
8644           Spicejet      2019-05-01   Chennai     Kolkata  09:45:00   
4601             Indigo      2019-05-01   Kolkata    Banglore  08:10:00   
8532        Jet Airways      2019-05-27     Delhi      Cochin  05:30:00   
2216        Jet Airways      2019-06-12    Mumbai   Hyderabad  19:35:00   
2480        Jet Airways      2019-06-12     Delhi      Cochin  14:00:00   

     arrival_time  duration  total_stops              additional_info  price  
5617     01:30:00       305          1.0                      No info   5054  
6228     22:30:00       765          1.0                      No info   9646  
6772     15:15:00       980          1.0  In-flight meal not included  11087  
6423     21:00:00       480          1.0                      No info  13587  
1639     04:25:00      1760          2.0                      No info  16704  
...           ...       ...          ...                          ...    ...  
8644     12:00:00       135          0.0                      No info   3597  
4601     13:00:00       290          1.0                      No info   5069  
8532     12:35:00       425          2.0  In-flight meal not included  15544  
2216     21:05:00        90          0.0  In-flight meal not included   3210  
2480     19:00:00       300          1.0                      No info  14714  

[1674 rows x 10 columns]